### PR TITLE
fix(live-codeblock): add error boundary to live code preview

### LINKED ISSUE #8009: `undefined` crashes live codeblock
### Have you read the Contributing Guidelines on issues?

- [X] I have read the [Contributing Guidelines on issues](https://github.com/facebook/docusaurus/blob/main/CONTRIBUTING.md#reporting-new-issues).

### Prerequisites

- [X] I'm using the latest version of Docusaurus.
- [ ] I have tried the `npm run clear` or `yarn clear` command.
- [ ] I have tried `rm -rf node_modules yarn.lock package-lock.json` and re-installing packages.
- [ ] I have tried creating a repro with https://new.docusaurus.io.
- [ ] I have read the console error message carefully (if applicable).

### Description

The live codeblock crashes the entire page on inputs that cause React to try to render `undefined`.

I can work around this by changing `transformCode` to have an `|| null` in case the code returns `undefined`.

### Reproducible demo

_No response_

### Steps to reproduce

* Open https://docusaurus.io/docs/next/markdown-features/code-blocks#interactive-code-editor
* Replace the text in one of the live code blocks with the text `undefined`

### Expected behavior

The entire page does not crash and the code block renders an error instead.

### Actual behavior

The entire page (not just the code block) crashes with the error message `Minified React error #152; visit https://reactjs.org/docs/error-decoder.html?invariant=152&args[]=n for the full message or use the non-minified dev environment for full errors and additional helpful warnings.`

### Your environment

- Public source code:
- Public site URL: https://docusaurus.io/docs/next/markdown-features/code-blocks#interactive-code-editor
- Docusaurus version used: latest on website and 2.0.1
- Environment name and version (e.g. Chrome 89, Node.js 16.4):
- Operating system and version (e.g. Ubuntu 20.04.2 LTS):

### Self-service

- [ ] I'd be willing to fix this bug myself.

### PR DESCRIPTION
**Pull request by rashidmya**
## Pre-flight checklist

- [x] I have read the [Contributing Guidelines on pull requests](https://github.com/facebook/docusaurus/blob/main/CONTRIBUTING.md#pull-requests).
- [ ] **If this is a code change**: I have written unit tests and/or added dogfooding pages to fully verify the new behavior.
- [ ] **If this is a new API or substantial change**: the PR has an accompanying issue (closes #0000) and the maintainers have approved on my working plan.

## Motivation
Fixes page crashing when typing `undefined` or `{}` in live editor

### Test links

Deploy preview: https://deploy-preview-8015--docusaurus-2.netlify.app/

## Related issues/PRs

fix #8009

### CODE REVIEW

**Comment by Josh-Cena** [comment]
Note, we already have an error boundary component in core: https://docusaurus.io/docs/docusaurus-core#errorboundary Please just reuse this one

**Comment by rashidmya** [comment]
using error boundary component in core results in this behavior inside the live editor. is that okay?

Capture

**Comment by Josh-Cena** [comment]
You should provide a custom `fallback`.

**Comment by rashidmya** [comment]
Its working correctly now. any notes?

**Review by slorber — CHANGES_REQUESTED** [review]
This does not render very nicely

CleanShot 2022-09-01 at 12 24 47@2x


Please make sure the error is displayed nicely, maybe keep the "results" header tab, and ensure there's still the padding/background?

We want the design to be consistent with other playground errors such as:

CleanShot 2022-09-01 at 12 28 46@2x

Maybe you could reuse the `LiveError` component? 🤷‍♂️ 

 Note you introduce a new translation label but do not provide translations for it (and thus CI fails). You could reuse the key "theme.ErrorPageContent.tryAgain" to fix this. I can't merge if the CI fails

**Comment by rashidmya** [comment]
Capture
is this better?

I had difficulty figuring out how to reuse the `LiveError` component

File: packages/docusaurus-theme-live-codeblock/src/theme/Playground/index.tsx [hunk 1]
```diff
@@ -81,6 +83,30 @@ function EditorWithHeader() {
   );
 }
 
+function ErrorFallbackWithHeader({error, tryAgain}: ErrorProps): JSX.Element {
```
Comment (slorber) [inline packages/docusaurus-theme-live-codeblock/src/theme/Playground/index.tsx:86]: too much duplication here

Why not putting the error boundary inside ResultWithHeader instead? just wrap LivePreview?

**Review by slorber — CHANGES_REQUESTED** [review]
this looks better yes!

File: packages/docusaurus-theme-live-codeblock/src/theme/Playground/index.tsx (hunk 1 above)
Comment (rashidmya) [reply packages/docusaurus-theme-live-codeblock/src/theme/Playground/index.tsx:86]: true

Comment (rashidmya) [reply packages/docusaurus-theme-live-codeblock/src/theme/Playground/index.tsx:86]: Capture

wrapping `LivePreview` gives this behavior

**Comment by rashidmya** [comment]
@slorber why does CI fail when I added styles for the component

**Comment by slorber** [comment]
> @slorber why does CI fail when I added styles for the component

The CI says it: you just added an extra `!` in the tryAgain button label, and we end up with 2 buttons in our theme with the same key but different labels which we should avoid.

LGTM 👍  just did minor fixes


---

Note I tried to do some kind of "auto try again on code change" behavior but it's complicated because it seems `const {code,error} = useContext(LiveContext)` does not update for some reason when code string contains error, so the auto retry is hard to trigger with an effect.

Also we still have weird annoying cases where both the live error and the error boundary crash are displayed at the same time. Tried to improve but react-live does not make it very easy unfortunately

When code change from "crash" to "react-live-error" state, both are displayed. Unless we find a way to autoretry the error boundary I think it's better to display the live error in addition (even if it's weird) rather that staying on the rough error boundary crash message

**Pull request merged**